test/route-rate-limit.test.js

```diff
@@ -572,3 +572,61 @@ test('custom error response', t => {
     })
   })
 })
+
+test('variable max contenders', t => {
+  t.plan(18)
+  const fastify = Fastify()
+  fastify.register(rateLimit, {
+    global: false,
+    max: 1,
+    timeWindow: 10000
+  })
+
+  fastify.get('/', {
+    config: {
+      rateLimit: {
+        keyGenerator (req) { return req.headers['api-key'] },
+        max: (req, key) => { return key === 'pro' ? 3 : 2 }
+      }
+    }
+  }, (req, res) => { res.send('hello') })
+
+  fastify.get('/limit', { config: { rateLimit: {} } }, (req, res) => { res.send('limited') })
+
+  fastify.inject({ url: '/', headers: { 'api-key': 'pro' } }, (err, res) => {
+    t.error(err)
+    t.strictEqual(res.statusCode, 200)
+    fastify.inject({ url: '/', headers: { 'api-key': 'pro' } }, (err, res) => {
+      t.error(err)
+      t.strictEqual(res.statusCode, 200)
+      fastify.inject({ url: '/', headers: { 'api-key': 'pro' } }, (err, res) => {
+        t.error(err)
+        t.strictEqual(res.statusCode, 200)
+        fastify.inject({ url: '/', headers: { 'api-key': 'pro' } }, (err, res) => {
+          t.error(err)
+          t.strictEqual(res.statusCode, 429)
+          fastify.inject({ url: '/limit', headers: { 'api-key': 'pro' } }, (err, res) => {
+            t.error(err)
+            t.strictEqual(res.statusCode, 200)
+            fastify.inject({ url: '/limit', headers: { 'api-key': 'pro' } }, (err, res) => {
+              t.error(err)
+              t.strictEqual(res.statusCode, 429, 'for route /limit the api-key must be ignored')
+              fastify.inject({ url: '/', headers: { 'api-key': 'not-pro' } }, (err, res) => {
+                t.error(err)
+                t.strictEqual(res.statusCode, 200)
+                fastify.inject({ url: '/', headers: { 'api-key': 'not-pro' } }, (err, res) => {
+                  t.error(err)
+                  t.strictEqual(res.statusCode, 200)
+                  fastify.inject({ url: '/', headers: { 'api-key': 'not-pro' } }, (err, res) => {
+                    t.error(err)
+                    t.strictEqual(res.statusCode, 429)
+                  })
+                })
+              })
+            })
+          })
+        })
+      })
+    })
+  })
```

Comment:
I would refactor this code a bit to avoid so much nesting.

Reply:
Done